File: src/BYODClient.py
from tkinter import simpledialog, messagebox
import socket
import ssl

class BYODClient:
    def __init__(self):

        protocols = ['TLSv1.3']

        # Create an SSL/TLS context and set the options
        context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
        context.load_verify_locations('certificate.pem')
        context.set_alpn_protocols(protocols)
        #context.check_hostname = False # Does not verify the server name

        # Create a TCP/IP socket
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

        with context.wrap_socket(sock, server_hostname='Unknown') as ssock: # The name that has been put in the server certificate must be entered
            ssock.connect(('localhost', 7070)) # ip and port of the server

            # To send message to the server
            output = ssock.makefile('w')

            username = simpledialog.askstring('Username', 'Introduzca su nombre de usuario:')
            output.write(username + '\n')

            password = simpledialog.askstring('Password', 'Introduzca su contraseña:')
            output.write(password + '\n')

            msg = simpledialog.askstring('Message', 'Introduzca su mensaje:')
            output.write(msg + '\n')

            output.flush()

            # Read the server response
            input_data = ssock.makefile('r')
            response = input_data.readline()

            messagebox.showinfo('Response', response)

            # The connection is automatically closed when leaving the "with" block
            
if __name__ == '__main__':
    BYODClient()

File: src/BYODClient.java
import java.io.*;
import javax.net.ssl.SSLSocket;
import javax.net.ssl.SSLSocketFactory;
import javax.swing.JOptionPane;

public class BYODClient {

    private static final String[] protocols = new String[] { "TLSv1.3" };
    private static final String[] cipher_suites = new String[] { "TLS_AES_128_GCM_SHA256", "TLS_AES_256_GCM_SHA384",
            "TLS_CHACHA20_POLY1305_SHA256" };

    public BYODClient() {
        try {
            SSLSocketFactory socketFactory = (SSLSocketFactory) SSLSocketFactory.getDefault();
            SSLSocket socket = (SSLSocket) socketFactory.createSocket("localhost", 7070);
            socket.setEnabledProtocols(protocols);
            socket.setEnabledCipherSuites(cipher_suites);

            // Creates a PrintWriter to send message/MAC to the server
            PrintWriter output = new PrintWriter(new OutputStreamWriter(socket.getOutputStream()));

            String username = JOptionPane.showInputDialog(null, "Introduzca su nombre de usuario:");
            output.println(username);

            String password = JOptionPane.showInputDialog(null, "Introduzca su contraseña:");
            output.println(password);

            String msg = JOptionPane.showInputDialog(null, "Introduzca su mensaje:");
            output.println(msg);

            // Important for the message to be sent
            output.flush();

            // Creates a BufferedReader object to read the response from the server
            BufferedReader input = new BufferedReader(new InputStreamReader(socket.getInputStream()));

            // Read the server response
            String response = input.readLine();

            // Displays the server response
            JOptionPane.showMessageDialog(null, response);

            // The connection is closed
            output.close();
            input.close();
            socket.close();
        } catch (IOException ioException) {
            ioException.printStackTrace();
        } finally {
            System.exit(0);
        }
    }

    public static void main(String args[]) {
        new BYODClient();
    }
}
